src/main/java/com/linecorp/armeria/client/SessionProtocolNegotiationCache.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.LruMap;
+
+/**
+ * Keeps the recent {@link SessionProtocol} negotiation failures.
+ */
+public final class SessionProtocolNegotiationCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(SessionProtocolNegotiationCache.class);
+
+    private static final Map<String, CacheEntry> cache = new LruMap<String, CacheEntry>(65536) {
+        private static final long serialVersionUID = -2506868886873712772L;
+
+        @Override
+        protected boolean removeEldestEntry(Entry<String, CacheEntry> eldest) {
+            final boolean remove = super.removeEldestEntry(eldest);
+            if (remove) {
+                logger.debug("Evicted: '{}' does not support ", eldest.getKey(), eldest.getValue());
+            }
+
+            return remove;
+        }
+    };
+
+    public static boolean isUnsupported(SocketAddress remoteAddress, SessionProtocol protocol) {
+        final String key = key(remoteAddress);
+        final CacheEntry e;
+        synchronized (cache) {
```

Comment:
It depends on the server right? An armeria client talking to an HTTP1 server with no Keep-Alive will have short-lived connections AFAIK. In principle, it also "feels wrong" blocking threads on reading data that's in practice more or less immutable throughout the lifetime of a server, though I agree that the event loop will only be running on a few threads so the contention will not be so bad.

Java8 StampedLock is not reentrant, so maybe it has less overhead than RWL and would be an alternative? Or we can just go with ConcurrentHashMap, since having more than 65K hostnames connected to by a server seems pretty impossible? Or we can go with a synchronized map since you're right, there are only a few threads and map reads should be fast, just want to make sure to think about it a little bit more :)